google/cloud/storage/client.py

```diff
@@ -135,32 +140,79 @@ def __init__(
         self._initial_client_options = client_options
         self._extra_headers = extra_headers
 
-        kw_args = {"client_info": client_info}
-
-        # `api_endpoint` should be only set by the user via `client_options`,
-        # or if the _get_storage_host() returns a non-default value (_is_emulator_set).
-        # `api_endpoint` plays an important role for mTLS, if it is not set,
-        # then mTLS logic will be applied to decide which endpoint will be used.
-        storage_host = _get_storage_host()
-        _is_emulator_set = storage_host != _DEFAULT_STORAGE_HOST
-        kw_args["api_endpoint"] = storage_host if _is_emulator_set else None
+        connection_kw_args = {"client_info": client_info}
 
         if client_options:
             if isinstance(client_options, dict):
                 client_options = google.api_core.client_options.from_dict(
                     client_options
                 )
-            if client_options.api_endpoint:
-                api_endpoint = client_options.api_endpoint
-                kw_args["api_endpoint"] = api_endpoint
+
+        if client_options and client_options.universe_domain:
+            self._universe_domain = client_options.universe_domain
+        else:
+            self._universe_domain = None
+
+        storage_emulator_override = _get_storage_emulator_override()
+        api_endpoint_override = _get_api_endpoint_override()
+
+        # Determine the api endpoint. The rules are as follows:
```

Comment:
Thanks for adding comments here. Could you share more about how/why this order is determined? 

What happens when a user specifies both `universe_domain` and `api_endpoint_override`? I read in one of (the many) docs that it should be possible for a user to specify the endpoint override in addition to the universe. I guess it’s alright as long as endpoint override takes precedence… is my understanding correct?

Reply:
Yes, it's explicitly ok if `universe_domain` and `api_endpoint_override` are both enabled. The latter takes precedence and there is no special casing - `universe_domain` is still set for all purposes other than determining the endpoint.

The order was determined by reading the TPC design doc very carefully.